File: Framework/CSharp/Framework/Framework/SmartConvert.cs
/*
 * 项目：Smartkernel.Framework
 * 作者：曹艳白
 * 邮箱：caoyanbai@139.com
 */
using System;
using System.Drawing;
using System.Globalization;
using System.Text;

namespace Smartkernel.Framework
{
    /// <summary>
    /// 转换
    /// </summary>
    public static class SmartConvert
    {
        /// <summary>
        /// 尝试转换（失败则返回默认值）-- 转换为Int32类型
        /// </summary>
        /// <param name="input">输入</param>
        /// <param name="defaultValue">默认值</param>
        /// <returns>结果</returns>
        public static int TryToInt32(object input, int defaultValue = 0)
        {
            try
            {
                return Convert.ToInt32(input);
            }
            catch
            {
                return defaultValue;
            }
        }

        /// <summary>
        /// 尝试转换（失败则返回默认值）-- 转换为Int16类型
        /// </summary>
        /// <param name="input">输入</param>
        /// <param name="defaultValue">默认值</param>
        /// <returns>结果</returns>
        public static Int16 TryToInt16(object input, Int16 defaultValue = 0)
        {
            try
            {
                return Convert.ToInt16(input);
            }
            catch
            {
                return defaultValue;
            }
        }

        /// <summary>
        /// 尝试转换（失败则返回默认值）-- 转换为Int64类型
        /// </summary>
        /// <param name="input">输入</param>
        /// <param name="defaultValue">默认值</param>
        /// <returns>结果</returns>
        public static Int64 TryToInt64(object input, Int64 defaultValue = 0)
        {
            try
            {
                return Convert.ToInt64(input);
            }
            catch
            {
                return defaultValue;
            }
        }

        /// <summary>
        /// 尝试转换（失败则返回默认值）-- 转换为 Decimal 类型
        /// </summary>
        /// <param name="input">输入</param>
        /// <param name="defaultValue">默认值</param>
        /// <returns>结果</returns>
        public static Decimal TryToDecimal(object input, Decimal defaultValue = 0)
        {
            try
            {
                return Convert.ToDecimal(input);
            }
            catch
            {
                return defaultValue;
            }
        }

        /// <summary>
        /// 尝试转换（失败则返回默认值）-- 转换为 Double 类型
        /// </summary>
        /// <param name="input">输入</param>
        /// <param name="defaultValue">默认值</param>
        /// <returns>结果</returns>
        public static Double TryToDouble(object input, Double defaultValue = 0.00)
        {
            try
            {
                return Convert.ToDouble(input);
            }
            catch
            {
                return defaultValue;
            }
        }

        /// <summary>
        /// 尝试转换（失败则返回默认值）-- 转换为 String 类型
        /// </summary>
        /// <param name="input">输入</param>
        /// <param name="defaultValue">默认值</param>
        /// <returns>结果</returns>
        public static String TryToString(object input, String defaultValue = "")
        {
            try
            {
                return Convert.ToString(input);
            }
            catch
            {
                return defaultValue;
            }
        }

        /// <summary>
        /// 尝试转换（失败则返回默认值）-- 转换为 Char 类型
        /// </summary>
        /// <param name="input">输入</param>
        /// <param name="defaultValue">默认值</param>
        /// <returns>结果</returns>
        public static Char TryToChar(object input, Char defaultValue = ' ')
        {
            try
            {
                return Convert.ToChar(input);
            }
            catch
            {
                return defaultValue;
            }
        }

        /// <summary>
        /// 尝试转换（失败则返回默认值）-- 转换为 DateTime类型
        /// </summary>
        /// <param name="input">输入</param>
        /// <param name="defaultValue">默认值</param>
        /// <returns>结果</returns>
        public static DateTime TryToDateTime(object input, DateTime defaultValue = default(DateTime))
        {
            try
            {
                return Convert.ToDateTime(input);
            }
            catch
            {
                return defaultValue;
            }
        }

        /// <summary>
        /// 尝试转换（失败则返回默认值）-- 转换为 bool 类型
        /// </summary>
        /// <param name="input">输入</param>
        /// <param name="defaultValue">默认值</param>
        /// <returns>结果</returns>
        public static bool TryToBool(object input, bool defaultValue = false)
        {
            try
            {
                return Convert.ToBoolean(input);
            }
            catch
            {
                return defaultValue;
            }
        }

        /// <summary>
        /// 尝试转换（失败则返回默认值）-- 转换为 byte 类型
        /// </summary>
        /// <param name="input">输入</param>
        /// <param name="defaultValue">默认值</param>
        /// <returns>结果</returns>
        public static byte TryToByte(object input, byte defaultValue = 0)
        {
            try
            {
                return Convert.ToByte(input);
            }
            catch
            {
                return defaultValue;
            }
        }

        /// <summary>
        /// 尝试转换（失败则返回默认值）-- 转换为 SByte 类型
        /// </summary>
        /// <param name="input">输入</param>
        /// <param name="defaultValue">默认值</param>
        /// <returns>结果</returns>
        public static SByte TryToSByte(object input, SByte defaultValue = 0)
        {
            try
            {
                return Convert.ToSByte(input);
            }
            catch
            {
                return defaultValue;
            }
        }

        /// <summary>
        /// 尝试转换（失败则返回默认值）-- 转换为 float 类型
        /// </summary>
        /// <param name="input">输入</param>
        /// <param name="defaultValue">默认值</param>
        /// <returns>结果</returns>
        public static float TryToSingle(object input, float defaultValue = 0)
        {
            try
            {
                return Convert.ToSingle(input);
            }
            catch
            {
                return defaultValue;
            }
        }

        /// <summary>
        /// 尝试转换（失败则返回默认值）-- 转换为 UInt32 类型
        /// </summary>
        /// <param name="input">输入</param>
        /// <param name="defaultValue">默认值</param>
        /// <returns>结果</returns>
        public static UInt32 TryToUInt32(object input, UInt32 defaultValue = 0)
        {
            try
            {
                return Convert.ToUInt32(input);
            }
            catch
            {
                return defaultValue;
            }
        }

        /// <summary>
        /// 尝试转换（失败则返回默认值）-- 转换为 UInt16 类型
        /// </summary>
        /// <param name="input">输入</param>
        /// <param name="defaultValue">默认值</param>
        /// <returns>结果</returns>
        public static UInt16 TryToUInt16(object input, UInt16 defaultValue = 0)
        {
            try
            {
                return Convert.ToUInt16(input);
            }
            catch
            {
                return defaultValue;
            }
        }

        /// <summary>
        /// 尝试转换（失败则返回默认值）-- 转换为 UInt64 类型
        /// </summary>
        /// <param name="input">输入</param>
        /// <param name="defaultValue">默认值</param>
        /// <returns>结果</returns>
        public static UInt64 TryToUInt64(object input, UInt64 defaultValue = 0)
        {
            try
            {
                return Convert.ToUInt64(input);
            }
            catch
            {
                return defaultValue;
            }
        }

        /// <summary>
        /// 从源类型转换为目标类型
        /// </summary>
        /// <param name="source">源</param>
        /// <param name="type">目标类型（可以是枚举、TimeSpan、DateTime、List[string]、Color、其他基本类型等）</param>
        /// <returns>转换之后的对象</returns>
        public static object To(object source, Type type)
        {
            object result = source;

            if (source == null)
            {
                return null;
            }

            string sourceString = source == null ? null : source.ToString();

            if (Nullable.GetUnderlyingType(type) != null)
            {
                //如果目标类型是Nullable类型，取原始类型
                type = Nullable.GetUnderlyingType(type);
            }

            if (type.IsEnum)
            {
                //目标类型为枚举类型
                result = Enum.Parse(type, sourceString, true);
            }
            else if (source is Enum && type == typeof(int))
            {
                //源为枚举类型，目标为整形
                result = (int)source;
            }
            else if (source is int && type.IsEnum)
            {
                //源为整形，目标为枚举类型
                result = Convert.ChangeType(source, type, CultureInfo.CurrentCulture);
            }
            else if (type == typeof(DateTime))
            {
                //目标类型为时间类型
                result = DateTime.Parse(sourceString);
            }
            else if (source is bool && type == typeof(string))
            {
                //如果源是布尔值，目标是字符串
                result = sourceString.ToLower();
            }
            else if (source is IConvertible)
            {
                //一般类型的转换
                result = Convert.ChangeType(source, type, CultureInfo.CurrentCulture);
            }

            return result;
        }

        /// <summary>
        /// 从源类型转换为目标类型
        /// </summary>
        /// <typeparam name="T">目标类型（可以是枚举、TimeSpan、DateTime、List[string]、Color、其他基本类型等）</typeparam>
        /// <param name="source">源</param>
        /// <returns>转换之后的对象</returns>
        public static T To<T>(object source)
        {
            return (T)To(source, typeof(T));
        }
    }
}

File: Framework/Java/framework/framework/smartkernel/framework/SmartConvert.java
/*
 * 项目：Smartkernel.Framework
 * 作者：曹艳白
 * 邮箱：caoyanbai@139.com
 */
package smartkernel.framework;

import java.lang.reflect.*;
import java.util.*;

/**
 * 转换
 * 
 */
public class SmartConvert {
	/**
	 * 尝试转换（失败则返回默认值）-- 转换为Int32类型
	 * 
	 * @param input
	 *            ，输入
	 * @param defaultValue
	 *            ，默认值
	 * @return，结果
	 */
	public static int tryToInt32(Object input, int defaultValue) {
		try {
			return Integer.parseInt(input.toString());
		} catch (Exception err) {
			return defaultValue;
		}
	}

	/**
	 * 尝试转换（失败则返回默认值）-- 转换为Int32类型
	 * 
	 * @param input
	 *            ，输入
	 * @return，结果
	 */
	public static int tryToInt32(Object input) {
		return tryToInt32(input, 0);
	}

	/**
	 * 尝试转换（失败则返回默认值）-- 转换为Int16类型
	 * 
	 * @param input
	 *            ，输入
	 * @param defaultValue
	 *            ，默认值
	 * @return，结果
	 */
	public static short tryToInt16(Object input, short defaultValue) {
		try {
			return Short.parseShort(input.toString());
		} catch (Exception err) {
			return defaultValue;
		}
	}

	/**
	 * 尝试转换（失败则返回默认值）-- 转换为Int16类型
	 * 
	 * @param input
	 *            ，输入
	 * @return，结果
	 */
	public static short tryToInt16(Object input) {
		return tryToInt16(input, (short) 0);
	}

	/**
	 * 尝试转换（失败则返回默认值）-- 转换为Int64类型
	 * 
	 * @param input
	 *            ，输入
	 * @param defaultValue
	 *            ，默认值
	 * @return，结果
	 */
	public static long tryToInt64(Object input, long defaultValue) {
		try {
			return Long.parseLong(input.toString());
		} catch (Exception err) {
			return defaultValue;
		}
	}

	/**
	 * 尝试转换（失败则返回默认值）-- 转换为Int64类型
	 * 
	 * @param input
	 *            ，输入
	 * @return，结果
	 */
	public static long tryToInt64(Object input) {
		return tryToInt64(input, (long) 0);
	}

	/**
	 * 尝试转换（失败则返回默认值）-- 转换为Double类型
	 * 
	 * @param input
	 *            ，输入
	 * @param defaultValue
	 *            ，默认值
	 * @return，结果
	 */
	public static double tryToDouble(Object input, double defaultValue) {
		try {
			return Double.parseDouble(input.toString());
		} catch (Exception err) {
			return defaultValue;
		}
	}

	/**
	 * 尝试转换（失败则返回默认值）-- 转换为Double类型
	 * 
	 * @param input
	 *            ，输入
	 * @return，结果
	 */
	public static double tryToDouble(Object input) {
		return tryToDouble(input, 0.0);
	}

	/**
	 * 尝试转换（失败则返回默认值）-- 转换为Float类型
	 * 
	 * @param input
	 *            ，输入
	 * @param defaultValue
	 *            ，默认值
	 * @return，结果
	 */
	public static float tryToFloat(Object input, float defaultValue) {
		try {
			return Float.parseFloat(input.toString());
		} catch (Exception err) {
			return defaultValue;
		}
	}

	/**
	 * 尝试转换（失败则返回默认值）-- 转换为Float类型
	 * 
	 * @param input
	 *            ，输入
	 * @return，结果
	 */
	public static float tryToFloat(Object input) {
		return tryToFloat(input, (float) 0.0);
	}

	/**
	 * 尝试转换（失败则返回默认值）-- 转换为 String 类型
	 * 
	 * @param input
	 *            ，输入
	 * @param defaultValue
	 *            ，默认值
	 * @return，结果
	 */
	public static String tryToString(Object input, String defaultValue) {
		try {
			return String.valueOf(input);
		} catch (Exception err) {
			return defaultValue;
		}
	}

	/**
	 * 尝试转换（失败则返回默认值）-- 转换为 String 类型
	 * 
	 * @param input
	 *            ，输入
	 * @return，结果
	 */
	public static String tryToString(Object input) {
		return tryToString(input, "");
	}

	/**
	 * 尝试转换（失败则返回默认值）-- 转换为 Char 类型
	 * 
	 * @param input
	 *            ，输入
	 * @param defaultValue
	 *            ，默认值
	 * @return，结果
	 */
	public static char tryToChar(Object input, char defaultValue) {
		try {
			return input.toString().charAt(0);
		} catch (Exception err) {
			return defaultValue;
		}
	}

	/**
	 * 尝试转换（失败则返回默认值）-- 转换为 Char 类型
	 * 
	 * @param input
	 *            ，输入
	 * @return，结果
	 */
	public static char tryToChar(Object input) {
		return tryToChar(input, ' ');
	}

	/**
	 * 尝试转换（失败则返回默认值）-- 转换为 DateTime类型
	 * 
	 * @param input
	 *            ，输入
	 * @param defaultValue
	 *            ，默认值
	 * @return，结果
	 */
	public static Date tryToDateTime(Object input, Date defaultValue) {
		try {
			return SmartDateTime.parse(input.toString());
		} catch (Exception err) {
			return defaultValue;
		}
	}

	/**
	 * 尝试转换（失败则返回默认值）-- 转换为 DateTime类型
	 * 
	 * @param input
	 *            ，输入
	 * @return，结果
	 */
	public static Date tryToDateTime(Object input) {
		return tryToDateTime(input, SmartDateTime.parse("1900-01-01"));
	}

	/**
	 * 尝试转换（失败则返回默认值）-- 转换为 bool 类型
	 * 
	 * @param input
	 *            ，输入
	 * @param defaultValue
	 *            ，默认值
	 * @return，结果
	 */
	public static boolean tryToBool(Object input, boolean defaultValue) {
		try {
			return Boolean.parseBoolean(input.toString());
		} catch (Exception err) {
			return defaultValue;
		}
	}

	/**
	 * 尝试转换（失败则返回默认值）-- 转换为 bool 类型
	 * 
	 * @param input
	 *            ，输入
	 * @return，结果
	 */
	public static boolean tryToBool(Object input) {
		return tryToBool(input, false);
	}

	/**
	 * 尝试转换（失败则返回默认值）-- 转换为 byte 类型
	 * 
	 * @param input
	 *            ，输入
	 * @param defaultValue
	 *            ，默认值
	 * @return，结果
	 */
	public static byte tryToByte(Object input, byte defaultValue) {
		try {
			return Byte.parseByte(input.toString());
		} catch (Exception err) {
			return defaultValue;
		}
	}

	/**
	 * 尝试转换（失败则返回默认值）-- 转换为 byte 类型
	 * 
	 * @param input
	 *            ，输入
	 * @return，结果
	 */
	public static byte tryToByte(Object input) {
		return tryToByte(input, (byte) 0);
	}

	/**
	 * 从源类型转换为目标类型
	 * 
	 * @param source
	 *            ，源
	 * @param type
	 *            ，目标类型（可以是枚举、TimeSpan、DateTime、List[string]、Color、其他基本类型等）
	 * @return，转换之后的对象
	 */
	@SuppressWarnings({ "unchecked", "rawtypes" })
	public static Object to(Object source, Class type) {
		Object result = source;

		if (source == null) {
			return null;
		}
		Class sourceType = source.getClass();
		String sourceString = source == null ? null : source.toString();

		if (sourceType.equals(type)) {
			return source;
		} else if (type.isEnum()) {
			// 目标类型为枚举类型
			result = SmartEnum.parse(type, sourceString);
		} else if (sourceType.isEnum() && type.equals(Integer.class)) {
			// 源为枚举类型，目标为整形
			result = Enum.valueOf(sourceType, sourceString).ordinal();
		} else if (sourceType.equals(Integer.class) && type.isEnum()) {
			// 源为整形，目标为枚举类型
			Iterator iterator = EnumSet.allOf(type).iterator();
			while (iterator.hasNext()) {
				Object current = iterator.next();

				String currentString = iterator.next().toString();
				int target = Enum.valueOf(type, currentString).ordinal();
				if ((int) source == target) {
					result = current;
				}
			}
		} else if (type.equals(Date.class)) {
			// 目标类型为时间类型
			result = SmartDateTime.parse(sourceString);
		} else if ((sourceType.equals(Boolean.class) || sourceType
				.equals(boolean.class)) && type.equals(String.class)) {
			// 如果源是布尔值，目标是字符串
			result = sourceString.toLowerCase();
		} else if ((type.equals(Boolean.class) || type.equals(boolean.class))
				&& sourceType.equals(String.class)) {
			// 如果源是字符串，目标是布尔值
			result = Boolean.valueOf(sourceString);
		} else if (SmartType.isNumbericType(sourceType)
				&& type.equals(String.class)) {
			// 如果源是数字类型，目标是字符串类型
			result = sourceString.toLowerCase();
		} else if (SmartType.isNumbericType(type)
				&& sourceType.equals(String.class)) {
			// 如果源是字符串类型，目标是数字类型
			if (type.equals(Integer.class) || type.equals(int.class)) {
				result = Integer.valueOf(sourceString.toLowerCase());
			} else if (type.equals(Long.class) || type.equals(long.class)) {
				result = Long.valueOf(sourceString.toLowerCase());
			} else if (type.equals(Float.class) || type.equals(float.class)) {
				result = Float.valueOf(sourceString.toLowerCase());
			} else if (type.equals(Double.class) || type.equals(double.class)) {
				result = Double.valueOf(sourceString.toLowerCase());
			} else if (type.equals(Short.class) || type.equals(short.class)) {
				result = Short.valueOf(sourceString.toLowerCase());
			} else if (type.equals(Byte.class) || type.equals(byte.class)) {
				result = Byte.valueOf(sourceString.toLowerCase());
			}
		} else if (sourceType.equals(String.class)
				&& (type.equals(Character.class) || type.equals(char.class))) {
			// 源为字符串，目标为字符
			return sourceString.charAt(0);
		} else if (type.equals(String.class)
				&& (sourceType.equals(Character.class) || sourceType
						.equals(char.class))) {
			// 源为字符，目标为字符串
			return String.valueOf(sourceString);
		} else {
			try {
				Constructor constructor = type.getConstructor(sourceType);
				result = constructor.newInstance(source);
			} catch (Exception err) {
			}
		}

		return result;
	}
}
